CIP-2551/CIP-2551.md

```diff
@@ -0,0 +1,271 @@
+---
+CIP: ?
+Title: Ed25519 elliptic curve group primitives
+Authors: Morgan Thomas <morgan.thomas@platonic.systems>
+Discussions-To: <morgan.thomas@platonic.systems>
+Status: Draft
+Type: Standards Track
+Created: 2022-07-27
+* License: Apache-2.0
+---
+
+## Preamble
+
+RFC 822 style headers containing metadata about the CIP, including the CIP
+number, a short descriptive title (limited to a maximum of 44 characters),
+the names, and optionally the contact info for each author, etc.
+
+## Simple Summary
+
+This CIP proposes that Plutus be extended with native code
+implementations of certain mathematical operations which are useful for
+cryptography. These mathematical operations are used in the implementation
+of the Ed25519 public/private key signing protocol. Cardano already
+supports a native code implementation of verifying signatures created by
+the Ed25519 signing protocol, but it does not enable Plutus to scripts
+perform the underlying mathematical operations directly. Being able to do
+this would be useful for implementing other cryptographic protocols based
+on the Ed25519 mathematical operations.  Applications include verifying
+zero-knowledge proofs, such as in the context of implementing zk-rollups.
+
+## Abstract
+
+Extend Plutus with primitives for performing calculations over the
+Ed25519 curve and the corresponding finite field.  These operations are
+to include the basic group and field operations, as well as multiscalar
+multiplication.
+
+## Motivation
+
+CIP-0381 proposes inclusion of the Bls12-381 curve in Plutus. Bls12-381
+is a pairing-friendly curve. Not all applications of elliptic curves
+require a pairing-friendly curve. Choosing a non-pairing-friendly
+curve where a pairing-friendly curve is not required can lead to better
+usage of computing power.  The Ed25519 curve allows for more efficient
+operations as compared to the Bls12-381 curve.  Pairing-friendly curves
+require the use of more bits in order to achieve security (255 bits
+for Ed25519 vs 381 bits for Bls12-381). Therefore it would be helpful,
+for those applications which do not require a pairing-friendly curve,
+to include a non-pairing-friendly curve in Plutus.
+
+An example application of the Ed25519 curve is in verifying
+zk-SNARKs on-chain. Halo 2, for example, can be used with the Ed25519
```

Comment:
How many EC operations are needed to implement Halo proof verification? This was a concern already with the Bls primitives: the primitives themselves are already expensive, and if we need many of them to actually perform Halo verification, then we may still not have enough resources to do so. It would be sad to implement these primitives and then find out we still couldn't do the thing we actually wanted to do.

I'd also be interested in knowing how long it takes to verify a typical Halo 2 proof on typical consumer hardware. If it's in the order of milliseconds then we are in trouble.